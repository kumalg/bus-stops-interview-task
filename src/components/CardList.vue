<template>
  <BaseCard :title="title" :without-paddings="true" class="card-list">
    <div v-if="filterable" class="search-row">
      <SearchInput v-model="userInput" class="search-input" />
    </div>
    <button
      v-if="reversible"
      class="list-title"
      @click="isReversed = !isReversed"
      :tabindex="reversible ? 0 : -1"
      type="button"
    >
      {{ subtitle }}
      <ArrowBottomIcon
        class="order-icon"
        :class="{ flip: isReversed }"
      />
    </button>
    <p v-else class="list-title">
      {{ subtitle }}
    </p>

    
    <transition name="fade" mode="out-in">
      <ul :key="`items_${filteredItems.length}_${isReversed}`" class="list">
        <li
          v-for="item in filteredItems"
          :key="item"
          class="list-item"
        >
          <button :tabindex="selectable ? 0 : -1" class="list-item-button" :class="{ selectable, active: selectable && modelValue === item }" type="button" @click="selectable && selectItem(item)">
            {{ item }}
          </button>
        </li>
      </ul>
    </transition>
  </BaseCard>
</template>

<script setup lang="ts">
import { computed, ref } from 'vue';
import ArrowBottomIcon from '@/icons/ArrowBottomIcon.vue';
import BaseCard from './BaseCard.vue';
import SearchInput from './SearchInput.vue';

const emit = defineEmits<{
  (e: 'update:model-value', item: string): void
}>()

const props = withDefaults(defineProps<{
  items: string[],
  subtitle: string,
  title?: string,
  reversible?: boolean,
  selectable?: boolean,
  filterable?: boolean,
  modelValue?: string | null,
}>(), {
  title: '',
  reversible: false,
  selectable: false,
  modelValue: null
})

const userInput = ref('')
const isReversed = ref(false)

const reversedItem = computed(() => {
  const copy = [...props.items]
  copy.reverse()
  return copy
})

const orderedItems = computed(() => {
  return isReversed.value ? reversedItem.value : props.items
})

const filteredItems = computed(() => {
  if (!props.filterable || !userInput.value) {
    return orderedItems.value
  }

  return orderedItems.value.filter(item => item.toLowerCase().includes(userInput.value.trim().toLowerCase()))
})

const selectItem = (item: string) => {
  emit('update:model-value', item)
}
</script>

<style lang="scss" scoped>
.card-list {
  .list-title {
    display: block;
    background-color: transparent;
    border: none;
    width: 100%;
    font-size: 0.75rem;
    line-height: 1rem;
    letter-spacing: 0.025em;
    font-weight: 600;
    display: flex;
    gap: 0.25rem;
    color: $color-gray-800;
    margin: 0;
    padding: $card-padding;
    border-bottom: 1px solid $color-gray-500;

    &:has(.order-icon) {
      cursor: pointer;

      &:hover .order-icon {
        color: $color-gray-800;
      }
    }

    .order-icon {
      color: $color-gray-700;

      @include default-transition('color, transform');

      &.flip {
        transform: rotate(180deg);
      }
    }
  }

  .search-row {
    padding: 0.5rem;

    .search-input {
      width: 18rem;
      max-width: 100%;
    }
  }

  .list {
    overflow: auto;
    list-style: none;
    padding: 0;
    margin: 0;
    min-height: 0;
  }

  .list-item {
    &:not(:last-child) {
      border-bottom: 1px solid $color-gray-400;
    }
  }

  .list-item-button {
    display: block;
    width: 100%;
    height: 3.5rem;
    border: none;
    background-color: transparent;
    text-align: left;
    padding: 0 $card-padding;
    color: $color-gray-800;
    font-family: inherit;
    user-select: none;

    @include default-transition('color, background-color');

    &.selectable {
      cursor: pointer;

      &:hover, &.hover {
        background-color: $color-gray-300;
      }

      &:active, &.active {
        color: $color-accent;
      }
    }
  }
}
</style>