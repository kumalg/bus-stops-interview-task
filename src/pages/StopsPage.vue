<template>
  <BaseCard>
    Stops
    <ul>
      <li v-for="stop in stops" :key="stop">{{  stop }}</li>
    </ul>
  </BaseCard>
</template>

<script setup lang="ts">
import BaseCard from '@/components/BaseCard.vue';
import { computed } from 'vue';
import { useStore } from 'vuex';

const store = useStore()

const stops = computed(() => store.getters.stops)
</script>

<style scoped>

</style>