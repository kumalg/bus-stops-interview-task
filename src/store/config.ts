import type { BusStop, BusStopsFetchStatus } from '@/types';

export enum StoreMutation {
    SetBusStops = 'SET_BUS_STOPS',
    SetStatus = 'SET_STATUS',
    SetLineStops = 'SET_LINE_STOPS',
    SetLineStopTimes = 'SET_LINE_STOP_TIMES'
}

export enum StoreAction {
    FetchBusStops = 'fetchBusStops',
    GetLineStops = 'getLineStops',
    GetTimesForLineStop = 'getTimesForLineStop'
}

export type LineStopsKey = BusStop['line'];
export type LineStopsValue = BusStop['stop'][];

export type LineStopTimesKey = {
    line: BusStop['line'];
    stop: BusStop['stop'];
};
export type LineStopTimesValue = BusStop['time'][];

export type StoreState = {
    busStops: BusStop[];
    status: BusStopsFetchStatus;
    lineStops: {
        [key in LineStopsKey]: LineStopsValue;
    };
    lineStopTimes: Map<LineStopTimesKey, LineStopTimesValue>;
};
