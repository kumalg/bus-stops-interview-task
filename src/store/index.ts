import type { BusStop } from '@/types';
import type { StoreState } from '@/store/config';

import { createStore, useStore as baseUseStore, Store } from 'vuex';
import { InjectionKey } from 'vue';

import { toSorted, toSortedTimes } from '@/assets/sort';
import { validateBusStops } from '@/assets/validator';

import { StoreAction, StoreMutation } from '@/store/config';

import { BusStopsFetchStatus } from '@/types';

export const key: InjectionKey<Store<StoreState>> = Symbol();

export const store = createStore<StoreState>({
    state: {
        busStops: [],
        status: BusStopsFetchStatus.UNFETCHED,
        lineStops: {},
        lineStopTimes: new Map()
    },
    getters: {
        lines: (state) => toSorted(new Set(state.busStops.map(({ line }) => line))),
        stops: (state) => toSorted(new Set(state.busStops.map(({ stop }) => stop)))
    },
    mutations: {
        [StoreMutation.SetBusStops](state, value: BusStop[]) {
            state.busStops = value;
        },
        [StoreMutation.SetStatus](state, value: BusStopsFetchStatus) {
            state.status = value;
        },
        [StoreMutation.SetLineStops](
            state,
            { line, stops }: { line: BusStop['line']; stops: BusStop['stop'][] }
        ) {
            state.lineStops[line] = stops;
        },
        [StoreMutation.SetLineStopTimes](
            state,
            {
                key,
                times
            }: { key: { line: BusStop['line']; stop: BusStop['stop'] }; times: BusStop['time'][] }
        ) {
            state.lineStopTimes.set(key, times);
        }
    },
    actions: {
        async [StoreAction.FetchBusStops]({ state, commit }) {
            if (state.status === BusStopsFetchStatus.FETCHING) {
                return;
            }

            commit(StoreMutation.SetStatus, BusStopsFetchStatus.FETCHING);

            try {
                const response = await fetch('http://localhost:3000/stops');

                if (!response.ok) {
                    commit(StoreMutation.SetStatus, BusStopsFetchStatus.ERROR);
                }

                const json = await response.json();

                if (!validateBusStops(json)) {
                    commit(StoreMutation.SetStatus, BusStopsFetchStatus.ERROR);
                } else {
                    commit(StoreMutation.SetBusStops, json);
                    commit(StoreMutation.SetStatus, BusStopsFetchStatus.FETCHED);
                }
            } catch {
                commit(StoreMutation.SetStatus, BusStopsFetchStatus.ERROR);
            }
        },
        [StoreAction.GetLineStops]({ state, commit }, line: number) {
            if (line in state.lineStops && typeof state.lineStops[line] === 'object') {
                return state.lineStops[line];
            }

            const stopFullObjects = state.busStops.filter((busStop) => busStop.line === line);
            stopFullObjects.sort(({ order: aOrder }, { order: bOrder }) => aOrder - bOrder);

            const stops = [...new Set(stopFullObjects.map(({ stop }) => stop))];

            commit(StoreMutation.SetLineStops, { line, stops });

            return stops;
        },
        [StoreAction.GetTimesForLineStop](
            { state, commit },
            key: { line: BusStop['line']; stop: BusStop['stop'] }
        ) {
            if (state.lineStopTimes.has(key)) {
                return state.lineStopTimes.get(key);
            }

            let times;
            try {
                times = toSortedTimes(
                    new Set(
                        state.busStops
                            .filter(({ line, stop }) => line === key.line && stop === key.stop)
                            .map(({ time }) => time)
                    )
                );
            } catch {
                return [];
            }

            commit(StoreMutation.SetLineStopTimes, { key, times });

            return times;
        }
    },
    modules: {}
});

export const useStore = () => {
    return baseUseStore(key);
};
