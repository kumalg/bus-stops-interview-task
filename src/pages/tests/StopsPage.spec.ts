import { shallowMount } from '@vue/test-utils';
import { createStore } from 'vuex';

import { key } from '@/store';

import StopsPage from '@/pages/StopsPage.vue';
import CardList from '@/components/CardList/CardList.vue';

describe('StopsPage', () => {
    const STOPS = ['stop1', 'stop2', 'stop3'];

    const StopsPageFactory = () => {
        const store = createStore({
            getters: {
                stops: () => STOPS
            }
        });
        return shallowMount(StopsPage, {
            global: {
                provide: {
                    [key as unknown as string]: store
                }
            }
        });
    };

    it('renders proper html markup', () => {
        const wrapper = StopsPageFactory();

        expect(wrapper.element).toMatchSnapshot();
    });

    it('renders correct stops list', () => {
        const wrapper = StopsPageFactory();

        expect(wrapper.findComponent(CardList).props()['items']).toEqual(STOPS);
    });
});
