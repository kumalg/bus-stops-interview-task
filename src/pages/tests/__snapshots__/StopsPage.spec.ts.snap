// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StopsPage renders proper html markup 1`] = `
<base-card-stub
  class="card-list stops-page"
  title=""
  withoutpaddings="true"
/>
`;
